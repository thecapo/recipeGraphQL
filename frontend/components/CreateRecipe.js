import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Router from "next/router";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { ALL_RECIPES_QUERY } from "./Recipes";

const CREATE_RECIPE_MUTATION = gql`
  mutation CREATE_RECIPE_MUTATION(
    $title: String!
    $ingredients: String!
    $instructions: String!
    $image: String
    $largeImage: String
  ) {
    createRecipe(
      title: $title
      ingredients: $ingredients
      instructions: $instructions
      image: $image
      largeImage: $largeImage
    ) {
      id
      title
    }
  }
`;

class CreateRecipe extends Component {
  state = {
    title: "",
    instructions: "",
    ingredients: "",
    image: "",
    largeImage: ""
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'recipeapp');
    const res = await fetch("https://api.cloudinary.com/v1_1/cawo/image/upload", {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    console.log(file)
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    })
  }

  render() {
    return (
      <Mutation
        mutation={CREATE_RECIPE_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: ALL_RECIPES_QUERY }]}
      >
        {(createRecipe, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createRecipe();
              Router.push({
                pathname: "/recipe",
                query: { id: res.data.createRecipe.id }
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  // required
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                  <img src={this.state.image} alt="Upload Preview" />
                )}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  defaultValue={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="ingredients">
                Ingredients
                <input
                  type="text"
                  id="ingredients"
                  name="ingredients"
                  placeholder="Ingredients..."
                  required
                  defaultValue={this.state.ingredients}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="instructions">
                Instructions
                <textarea
                  id="instructions"
                  name="instructions"
                  placeholder="Instructions..."
                  required
                  defaultValue={this.state.instructions}
                  onChange={this.handleChange}
                />
              </label>
              <button type="sumbit">Create</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateRecipe;
export { CREATE_RECIPE_MUTATION };