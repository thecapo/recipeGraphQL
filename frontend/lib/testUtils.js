import casual from 'casual';

casual.seed(777);

const fakeRecipe = () => ({
  __typename: "Recipe",
  id: "abc123",
  title: "Dessert Recipe",
  ingredients: "1/3tsbp ice cream, 1cup lollipop",
  instructions: "the quick brown fox jumps over the lazy dog.",
  image: "recipe.jpg",
  largeImage: "recipe-large.jpg",
  user: null,
  updatedAt: casual.date((format = "YYYY-MM-DD")),
  createdAt: casual.date((format = "YYYY-MM-DD"))
});

const fakeUser = () => ({ 
  __typename: "User",
  id: "1234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  updatedAt: casual.date((format = "YYYY-MM-DD")),
  createdAt: casual.date((format = "YYYY-MM-DD"))
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export default {
  fakeRecipe,
  fakeUser,
  LocalStorageMock
};