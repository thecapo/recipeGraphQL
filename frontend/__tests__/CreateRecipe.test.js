import { mount } from "enzyme";
import wait from "waait";
import toJSON from "enzyme-to-json";
import Router from "next/router";
import { MockedProvider } from "react-apollo/test-utils";
import CreateRecipe, { CREATE_RECIPE_MUTATION } from "../components/CreateRecipe";
import { fakeRecipe } from '../lib/testUtils';

const recipeImage = "https://recipe.com/recipe.jpg";

global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: recipeImage,
    eager: [{ secure_url: recipeImage }]
  })
});

describe("<CreateRecipe/>", () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateRecipe />
      </MockedProvider>
    )
    const form = wrapper.find('form[data-test="form"]')
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateRecipe />
      </MockedProvider>
    )
    const input = wrapper.find('input[type="file"]')
    input.simulate('change', { target: { files: ['fakeRecipe.jpg'] } })
    await wait();
    const component = wrapper.find('CreateRecipe').instance()
    expect(component.state.image)
  });
});
