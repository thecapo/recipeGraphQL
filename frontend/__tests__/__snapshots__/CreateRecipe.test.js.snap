// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateRecipe/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1kmq3n-0 kiKqiK"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        defaultValue=""
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
      />
    </label>
    <label
      htmlFor="ingredients"
    >
      Ingredients
      <input
        defaultValue=""
        id="ingredients"
        name="ingredients"
        onChange={[Function]}
        placeholder="Ingredients..."
        required={true}
        type="text"
      />
    </label>
    <label
      htmlFor="instructions"
    >
      Instructions
      <textarea
        defaultValue=""
        id="instructions"
        name="instructions"
        onChange={[Function]}
        placeholder="Instructions..."
        required={true}
      />
    </label>
    <button
      type="sumbit"
    >
      Create
    </button>
  </fieldset>
</form>
`;
